File: src/graphs/python/bellman_ford/src/bellman_ford.py
"""Bellman-Ford algorithm for shortest paths with negative edges."""

from __future__ import annotations

import math

from graph import Graph, Vertex


def bellman_ford(graph: Graph, source: Vertex, destination: Vertex) -> float:
    """
    Find the shortest path between source and destination using Bellman-Ford algorithm.

    This algorithm handles negative edge weights and detects negative cycles.

    Args:
        graph: The graph to search.
        source: The starting vertex.
        destination: The target vertex.

    Returns:
        The shortest distance from source to destination, or infinity if no path exists
        or if a negative cycle is detected.
    """
    if not (graph.contains(source) and graph.contains(destination)):
        return math.inf

    if source == destination:
        return 0

    distances: dict[Vertex, float] = {}

    for vertex in graph.vertices():
        distances[vertex] = math.inf

    distances[source] = 0

    for _ in range(graph.size() - 1):
        for edge in graph.edges():
            new_distance = distances[edge.source] + edge.distance

            if (
                distances[edge.source] != math.inf
                and new_distance < distances[edge.destination]
            ):
                distances[edge.destination] = new_distance

    for edge in graph.edges():
        if (
            distances[edge.source] != math.inf
            and distances[edge.source] + edge.distance < distances[edge.destination]
        ):
            # Graph contains negative weight cycle
            return math.inf

    return distances[destination]

File: src/graphs/cpp/bellman_ford/src/bellman_ford.cpp
#include "bellman_ford.h"
#include <climits>
#include <limits>
#include <unordered_map>

// Bellman-Ford algorithm: handles negative edge weights and detects negative cycles
template <class T>
int bellmanFord(const Graph<T> &graph, Vertex<T> source,
                Vertex<T> destination) {
  if (!graph.contains(source) || !graph.contains(destination))
    return INT_MAX;

  if (source == destination)
    return 0;

  std::unordered_map<Vertex<T>, int, HashFunction<T>> distances;

  for (const auto &vertex : graph.vertices()) {
    distances[vertex] = std::numeric_limits<int>::max();
  }

  distances[source] = 0;

  const auto allEdges = graph.edges();
  const size_t numVertices = graph.size();

  // Relax all edges V-1 times
  for (size_t i = 0; i < numVertices - 1; ++i) {
    bool updated = false;

    for (const auto &edge : allEdges) {
      if (distances[edge.source] != std::numeric_limits<int>::max()) {
        int newDistance = distances[edge.source] + edge.distance;
        if (newDistance < distances[edge.destination]) {
          distances[edge.destination] = newDistance;
          updated = true;
        }
      }
    }

    // Early termination if no updates were made
    if (!updated)
      break;
  }

  // Check for negative-weight cycles
  for (const auto &edge : allEdges) {
    if (distances[edge.source] != std::numeric_limits<int>::max() &&
        distances[edge.source] + edge.distance < distances[edge.destination]) {
      // Negative cycle detected
      return INT_MAX;
    }
  }

  return distances[destination] == std::numeric_limits<int>::max()
             ? INT_MAX
             : distances[destination];
}

template int bellmanFord<int>(const Graph<int> &graph, Vertex<int> source,
                              Vertex<int> destination);
template int bellmanFord<char>(const Graph<char> &graph, Vertex<char> source,
                               Vertex<char> destination);
template int bellmanFord<float>(const Graph<float> &graph, Vertex<float> source,
                                Vertex<float> destination);
template int bellmanFord<double>(const Graph<double> &graph,
                                 Vertex<double> source,
                                 Vertex<double> destination);
